### DISCUSSION TITLE
custom action before service start

### DISCUSSION THREAD

**Discussion by Tractorboy99** [wixtoolset/issues#7487] [Questions]
I am trying to install a service but need that app.config file to be editted before the service starts

so the order of actions should be
copy files
edit app.config
start up the service

I have the customactions written and the config updates, but only after the service has started, or the files are copied (ie code runs but does nothing)

I have googled and that suggests using After="InstallFiles" but the fires before the files are copied.  I have also used InstallFinalize but that fires after the service has started

I'd be keen to have a link to the order of actions, but also an idea of how to teak the wxs file to meet my requirements.

I have also found that the customtactions fire when the uninstall process runs.  This is completely unnecessary, so it would be good to have an idea of how to stop customactions firing during the uninstall process.

So my code stripped for simplicity

<InstallExecuteSequence>
   <RemoveExistingProducts Before="InstallInitialize"
  <Custom Action="DOCustomAction" After="InstallFiles" />
</InstallExecuteSequence
....
<Feature Id="Feature" Title="Title" Level="1">
<ComponentGroupRef Id="InstallFiles" />
<ComponentGroupRef Id="configureFiles" />
<ComponentGroupRef Id="InstallService" />

**Comment by BMurri** [comment]
I would suggest you want your custom action that edits the `app.config` file to be scheduled `Before="StartServices"` but your custom action must be deferred to accomplish what you are attempting.

For suppressing on uninstalls, you should add a condition. `NOT Installed` may be close to what you are looking for (it will only run when performing initial installations and major upgrades, but in no other scenarios). There are better options, such as running whenever your `app.config` file is written by MSI, but that depends on that file being the key path of it's component (you will want to use the component action variable for that component, consult the Windows Installer documentation).
